### ISSUE TITLE
Generalize architecture to have tasks with priorities

### ISSUE THREAD

**Issue by costrouc** [Quansight/conda-store#22]
We need tasks with priorities since deletion of environments could take awhile.

**Comment by costrouc** [comment]
This is a blocker for a few issues need to have a reasonable ways to handling tasks. I get the argument for using a formal message queue for this but I think that a database can work well enough in our case.

**Comment by costrouc** [comment]
Did a lot of research around this area. Here were the constraints:
 - I don't want to add yet another piece of infrastructure for conda-store to require (e.g. message queues... yes I know how great it would be to have that!) or to depend on a specific database with task locking behaviour e.g. Postgres SKIP LOCKED. So the initial solution shouldn't require a message queue but should support one if needed.
 - Support retrying a task a configurable number of times
 - Be resilient to worker/scheduler failure e.g. in the middle of building a conda environment and it runs out of memory and the worker dies (yes this happens in conda builds)
 - Support periodic tasks e.g. garbage collection

Here were the frameworks that I found somewhat fit the bill:
  - [apscheduler](https://apscheduler.readthedocs.io/en/stable/) they are working on a 4.0 release which is starting to look a lot like celery but currently does not support retrying tasks
  - [huey](https://github.com/coleifer/huey) but only supports sqlite not sqlalchemy
  - [celery](https://docs.celeryproject.org/en/stable/getting-started/introduction.html) what a tried and true tool that fits all of the requirements above

Currently it looks like celery is the clear choice and it is a battle tested package that could support message queues in the future.

**Comment by costrouc** [comment]
I did the most simple celery example and I think this will work well.

`app.py`

```python
from celery import Celery

CELERY_TASK_LIST = [
    'tasks',
]

app = Celery(
    'tasks',
    backend='db+sqlite:///celery.sqlite',
    broker='sqla+sqlite:///celery.sqlite',
    include=CELERY_TASK_LIST,
)

if __name__ == "__main__":
    app.start()
```

`tasks.py`

```python
from app import app

@app.task
def task_hello_world(message="World"):
    return f"Hello {message}"

@app.task
def task_addition(a, b):
    return a + b

@app.task
def task_multiply(a, b):
    return a * b
```

Then run a worker

`celery -A app worker -l INFO`

Finally trigger a task

```python
import tasks

result = tasks.task_multiply.delay(1, 2)
print(result.ready())
print(result.get(timeout=2))
```

**Comment by costrouc** [comment]
Closed by #90
